situp-plugins/service-map-stateful/src/main/java/com/amazon/situp/plugins/processor/ServiceMapRelationship.java

```diff
@@ -110,20 +129,95 @@ public void setTraceGroupName(String traceGroupName) {
         this.traceGroupName = traceGroupName;
     }
 
+    public String getHashId() {
+        return hashId;
+    }
+
+    public void setHashId(String hashId) {
+        this.hashId = hashId;
+    }
+
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         ServiceMapRelationship that = (ServiceMapRelationship) o;
         return Objects.equals(serviceName, that.serviceName) &&
-                Objects.equals(spanKind, that.spanKind) &&
+                Objects.equals(kind, that.kind) &&
                 Objects.equals(destination, that.destination) &&
                 Objects.equals(target, that.target) &&
                 Objects.equals(traceGroupName, that.traceGroupName);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(serviceName, spanKind, destination, target, traceGroupName);
+        return Objects.hash(serviceName, kind, destination, target, traceGroupName);
+    }
+
+    private String unhashedString() {
+        String result = serviceName + "," + kind + "," + traceGroupName + ",";
+        if (target != null) {
+            result += target.resource + "," + target.domain;
+        }
+        result += ",";
+        if(destination != null) {
+            result += destination.resource + "," + destination.domain;
+        }
+        return result;
+    }
+
+    private String md5Hash() {
+        if(THREAD_LOCAL_MESSAGE_DIGEST.get() == null) {
+            try {
+                THREAD_LOCAL_MESSAGE_DIGEST.set(MessageDigest.getInstance(MD5));
```

Comment:
Just curious. Is the usage of ThreadLocal to avoid thread competition on access to the "digest"(I mean the buffer) in MessageDigest?

Reply:
Right. Its to avoid multiple threads using the "update" function at the same time and potentially merging two of the hashes. The other option was to lock this object, but its not an expensive object